python/lsst/afw/multiband.py

```diff
@@ -0,0 +1,463 @@
+
+    @property
+    def origin(self):
+        """Minimum (y,x) position
+
+        This is the position of `self.getBBox().getMin()`,
+        but available as a tuple for numpy array indexing.
+        """
+        return np.array([self.y0, self.x0])
+
+    @property
+    def width(self):
+        """Width of the images
+        """
+        return self.getBBox().getWidth()
+
+    @property
+    def height(self):
+        """Height of the images
+        """
+        return self.getBBox().getHeight()
+
+    def __len__(self):
+        return len(self.filters)
+
+    def __getitem__(self, args):
+        """Get a slice of the underlying array
+
+        If only a single filter is specified,
+        return the single band object sliced
+        appropriately.
+        """
+        if not isinstance(args, tuple):
+            indices = (args,)
+        else:
+            indices = args
+
+        # Return the single band object if the first
+        # index is not a list or slice.
+        filters, filterIndex = self.filterToIndex(indices[0])
+        if not isinstance(filterIndex, slice) and len(filterIndex) == 1:
+            single = self.singles[filterIndex[0]]
+            # This temporary code is needed until image-like objects
+            # take integer indices, numpy-like slices
+            # and Point2I objects
+            # Once the image-like API has been updated,
+            # this entire "if" block can be replaced with:
+            # return self.singles[filterIndex[0]][indices[1:]]
+            if len(indices) > 1:
+                indexY, indexX = self.imageIndicesToNumpy(indices[1:])
+                # Return a scalar if possible
+                if isinstance(indices[1], Point2I) or (
+                    np.issubdtype(type(indexY), np.integer) and
+                    np.issubdtype(type(indexX), np.integer)
+                ):
+                    return self._slice(filters=filters, filterIndex=filterIndex, indices=indices[1:])
+                elif not isinstance(indices[1], Box2I):
+                    # Convert indices into a bounding box
+                    bbox = self.getBBoxFromIndices((indexY, indexX))
+                else:
+                    bbox = indices[1]
+                result = single.Factory(single, bbox, PARENT)
+            else:
+                result = single
+            return result
+
+        return self._slice(filters=filters, filterIndex=filterIndex, indices=indices[1:])
+
+    def filterToIndex(self, filterIndex):
+        """Convert a string of filter names to an index or a slice
+
+        Parameters
+        ----------
+        filterIndex: string, slice, or integer
+            Index to specify a filter or list of filters
+
+        Returns
+        -------
+        filters: list of `str`
+            Names of the filters in the slice
+        index: slice, int, or list of int's
+            Index to slice the parent with the
+            chosen filters.
```

Comment:
Ok, I fixed the docs and the code to (I think) make this clear:
```
        Convert a string of filter names to an index or a slice

        Because `filterIndex` can either be a string (or list of strings),
        an integer (or list of integers), or a slice, it can be useful to have
        both the names of all indices used (`filterNames`)
        and a slice or list of numerical indices to each filter name in
        `self.filters` (`filterIndices`).

        Parameters
        ----------
        filterIndex: string, slice, or integer
            Index to specify a filter or list of filters,
            for example "R" or ["R", "G", "B"] or 0 or [0,1]
            or slice(1, 3, None).

        Returns
        -------
        filterNames: list of `str`
            Names of the filters in the slice
        filterIndex: slice, int, or list of int's
            Index of each filter in `filterNames` in
            `self.filters`.
```

Reply:
When `filterIndex` is an integer or slice, does it still refer to filters?  I'm a bit worried that supporting both strings and integers for the same dimension could be confusing.  Do you have a use case for supporting integer indices for filters?

Reply:
Slicing. For example I might want `mImage[1:4]` to extract the image in 3 bands to build an RGB image.

This was more important before we changed the ordering of `__getitem__` to be `[x,y]` so that one could use the same `(filter, y, x)` index for both `mImage.array` and `mImage`, but might be less useful now if we implement a slicing method for eg. `mImage["R":"B"]`.

Another benefit is that if we eliminate numerical indices then `filters` wouldn't have to be a list of strings but could also be a list of `enum` types, which might be more robust.

Reply:
> `mImage["R":"B"]`

I'm not certain that syntax will actually work, but if it does, I do like that better.  (It looks like `slice("a", "b")` is totally valid; I haven't tried to see if that means `["a":"b"]` is).

> Another benefit is that if we eliminate numerical indices then filters wouldn't have to be a list of strings but could also be a list of enum types, which might be more robust.

I don't think we can go quite that far - I think it would get in the way of @isullivan's use case, and it's hard to do a compile- or input-time predefinition of all of the filters one might want to use.

Reply:
Ah, that makes sense.  So there will be no change in behavior based on the type of the first argument to `__getitem__`, as that first item will always be assumed to be the filter, and it's up to the user to make sure that type matches whatever was used to initialized the Multiband* object.

Reply:
right